protocol/strict-derivation.md

```diff
@@ -0,0 +1,355 @@
+staging channel, and the batch queue should always be empty right after the pipeline origin
+progresses.
+
+# Steady Batch Derivation
+
+Steady Batch Derivation changes how invalid batches and payload attributes at different stages in
+the derivation pipeline are treated.
+- Batch queue: if a batch is found to be invalid at the batch queue stage (`BatchDrop`), it is immediately replaced by an empty
+  batch, leading to deposit-only (first batch of an epoch) or empty payload attributes.
+- Engine queue: if the payload attributes are found by the EL client to be
+  [`INVALID`](https://github.com/ethereum/go-ethereum/blob/7fd7c1f7dd9ba8d90399df2f080e4101ae37a255/beacon/engine/errors.go#L63)
+  they are replaced by deposit-only/empty payload attributes and reinserted into the EL.
+
+## L1 origin selection
+
+There's an open design space about how to select the L1 origin when deriving one or more
+deposit-only blocks. I see 3 sensible options:
+
+- Option 1: Eagerly advance the L1 origin. This solves for missing L1 slots and will implicitly and
+automatically maintain a good L1/L2 block ratio.
+  - This guaranteed fast deposit inclusion for users and is similar to how a live sequencer
+  operates.
+  - Option 1.a: is a variation of this, where an in-protocol validation depth is maintained, so that
+  there's a constant sequencer drift, e.g. of 4 L1 blocks. This is even more similar to how a live
+  sequencer operates. This has the benefit that there's minimal disruption to the way that L1
+  origins change when going in and out of auto-derivation.
+  - Note that both, Options 1 and 1.a, wouldn't really suffer from shallow L1 reorgs, because only
+  invalid batches coming from L1 can cause this behavior, so if derivation is at the point where it
+  auto-derives blocks, a certain L1 inclusion block validation depth is already maintained.
+- Option 2: look at the last L1/L2-block-time-ratio-many (6 on mainnet) blocks and generate blocks
+in a way to maintain a steady L1/L2 block ratio, e.g. bumping the L1 origin selection every 6 blocks
+in the case of mainnet.
+  - This is similar to Option 1, but rather maintains the current sequencer drift, instead of
+  possibly more eagerly advancing it to get closer to the latest L1 origin.
+  - Edge case: the safe head is already very near the sequencer drift limit, and a newer origin
+  needs to be selected faster. This complicates this solution.
+  - Edge case: L1 missed a slot, and the L1 origin cannot advance as expected. This also needs to be
+  taken care of in Option 1.
+- Option 3: keep last L1 origin as long as possible, so until hitting the sequencing window. This is
+the simplest option, but also means that a sequencer has to fast-advance the L1 origin to catch up
+again. And if the L1 origin already slipped out of the drift window, a sequencer would need to
+continue creating deposit-only blocks until inside the drift window again.
```

Comment:
In the current implementation, it would actually already advance to keep within the drift window.